controllers/database_secret_handler.go

```diff
@@ -0,0 +1,217 @@
+}
+
+/* ------ Event Filter Functions ------ */
+
+func getDatabasesBySecret(c client.Client, secret types.NamespacedName) ([]kciv1alpha1.Database, []string, error) {
+	databaseList := kciv1alpha1.DatabaseList{}
+	if err := c.List(context.Background(), &databaseList, &client.ListOptions{Namespace: secret.Namespace}); err != nil {
+		return nil, nil, err
+	}
+	var matchedDbs []kciv1alpha1.Database
+	var matchedDbNames []string
+	for _, database := range databaseList.Items {
+		if database.Spec.SecretName == secret.Name {
+			matchedDbs = append(matchedDbs, database)
+			matchedDbNames = append(matchedDbNames, database.Name)
+		}
+	}
+	return matchedDbs, matchedDbNames, nil
+}
+
+func isFieldUpdated(dataOld map[string][]byte, dataNew map[string][]byte, keyName string) bool {
+
+	// read old and new value
+	oldValue, oldValueOk := dataOld[keyName]
+	newValue, newValueOk := dataNew[keyName]
+
+	if !oldValueOk || !newValueOk {
+		return false // values empty or do not exist
+	}
+
+	result := bytes.Compare(oldValue, newValue)
+	return result != 0 // true if values are not empty and updated
+}
+
+func isWatchedNamespace(watchNamespaces []string, ro runtime.Object) bool {
+	if watchNamespaces[0] == "" { // # it's necessary to set "" to watch cluster wide
+		return true // watch for all namespaces
+	}
+	// define object's namespace
+	objectNamespace := ""
+	database, isDatabase := ro.(*kciv1alpha1.Database)
+	if isDatabase {
+		objectNamespace = database.Namespace
+	} else {
+		secret, isSecret := ro.(*corev1.Secret)
+		if isSecret {
+			objectNamespace = secret.Namespace
+		} else {
+			logrus.Info("unknown object", "object", ro)
+			return false
+		}
+	}
+
+	// check that current namespace is watched by db-operator
+	for _, ns := range watchNamespaces {
+		if ns == objectNamespace {
+			return true
+		}
+	}
+	return false
+}
+
+func isDatabase(ro runtime.Object) bool {
+	_, isDatabase := ro.(*kciv1alpha1.Database)
+	return isDatabase
+}
+
+func isObjectUpdated(e event.UpdateEvent) bool {
+	if e.ObjectOld == nil {
+		logrus.Error(nil, "Update event has no old runtime object to update", "event", e)
+		return false
+	}
+	if e.ObjectNew == nil {
+		logrus.Error(nil, "Update event has no new runtime object for update", "event", e)
+		return false
+	}
+	// if object kind is a Database check that 'metadata.generation' field ('spec' section) has been changed
+	_, isDatabase := e.ObjectNew.(*kciv1alpha1.Database)
+	if isDatabase {
+		return e.ObjectNew.GetGeneration() != e.ObjectOld.GetGeneration()
+	}
+
+	// if object kind is a Secret check that password value has changed
+	secretNew, isSecret := e.ObjectNew.(*corev1.Secret)
+	if isSecret {
+		// if object kind is a Secret let's check that DB password inside has been updated
+		secretOld, secretOldOk := e.ObjectOld.(*corev1.Secret)
+		if !secretOldOk {
+			logrus.Error(nil, "Update Event error! ObjectOld is NOT a Secret: e.ObjectOld.GetName=", e.ObjectOld.GetName())
+			return false
+		}
+
```

Comment:
As I know, it is not best practice to use processing logic here, because "func [isObjectUpdated(e event.UpdateEvent)](https://github.com/vladimir22/db-operator/blob/a1bae504972338a86028a6e7b255ed9527fb173a/controllers/database_secret_handler.go#L169)  bool" used by `DatabaseReconciler` [eventFilter](https://github.com/vladimir22/db-operator/blob/a1bae504972338a86028a6e7b255ed9527fb173a/controllers/database_controller.go#L265) to filter out all `Secret` and `Database` events that need to be reconciled.


The "func (e *secretEventHandler) [Update](https://github.com/vladimir22/db-operator/blob/a1bae504972338a86028a6e7b255ed9527fb173a/controllers/database_secret_handler.go#L46)(evt event.UpdateEvent, q workqueue.RateLimitingInterface)"  (lines 59~68) contains logic to process already filtered out events.

Reply:
I understand. But I am not really happy about filtering with static key. Other solution I could think of is that adding annotation to the secrets used by db-operator and filtering with that. What do you think?

Reply:
Was thinking to add in the `createDatabase` a step to add a annotation for all secrets the db-operator is touching/using and only then check for this annotation here. Would remove the need for get here the DB engine.

Reply:
Agree, adding custom annotation into the database secret simplifies implementation.
@hyunysmile  @dabde  please have a look at my [implemetation](https://github.com/kloeckner-i/db-operator/pull/130/commits/31647d1f5646bd7f65978b0fd3baa285c053961a)
Thanks.